src/lem/interpreter.rs

```diff
@@ -1,180 +1,260 @@
 use crate::field::{FWrap, LurkField};
-use anyhow::{anyhow, bail, Result};
+use anyhow::{bail, Result};
 use std::collections::HashMap;
 
-use super::{pointers::Ptr, store::Store, symbol::Symbol, tag::Tag, Witness, LEM, LEMOP};
+use super::{
+    circuit::SlotsInfo, pointers::Ptr, store::Store, symbol::Symbol, tag::Tag, MetaPtr, LEM, LEMOP,
+};
 
-fn insert_into_ptrs<F: LurkField>(
-    ptrs: &mut HashMap<String, Ptr<F>>,
-    key: String,
-    value: Ptr<F>,
-) -> Result<()> {
-    let mut msg = "Key already defined: ".to_owned();
-    msg.push_str(&key);
-    if ptrs.insert(key, value).is_some() {
-        bail!("{msg}");
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub enum SlotType {
+    Hash2,
+    Hash3,
+    Hash4,
+}
+
+impl SlotType {
+    pub(crate) fn preimg_size(&self) -> usize {
+        match self {
+            Self::Hash2 => 4,
+            Self::Hash3 => 6,
+            Self::Hash4 => 8,
+        }
+    }
+}
+
+impl std::fmt::Display for SlotType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Hash2 => write!(f, "Hash2"),
+            Self::Hash3 => write!(f, "Hash3"),
+            Self::Hash4 => write!(f, "Hash4"),
+        }
     }
-    Ok(())
+}
+
+/// A `Slot` is characterized by an index and a type
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Slot {
+    pub idx: usize,
+    pub typ: SlotType,
+}
+
+impl std::fmt::Display for Slot {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Slot({}, {})", self.idx, self.typ)
+    }
+}
+
+/// This hashmap is populated during interpretation, telling which slots were
+/// visited and the data that was collected for each of them. The example below
+/// has 5 slots, 3 of which have been visited during interpretation.
+///
+///```text
+///            Slot index
+///      ┌┬┬┬┬┬───┬───┬───┐
+///      ├┼┼┼┼┤ 0 │ 1 │ 2 │
+///      ├┴┴┴┴┼───┼───┼───┤
+/// Slot │ H2 │ a │ b │   │
+/// type ├────┼───┼───┼───┘
+///      │ H3 │ c │   │
+///      └────┴───┴───┘
+///```
+/// `a`, `b` and `c` are the `Vec<Ptr<F>>` that were collected. The slots that
+/// weren't visited don't have key/value pairs present on `Visits`.
+pub(crate) type Visits<F> = HashMap<Slot, Vec<Ptr<F>>>;
+
+/// A `Frame` carries the data that results from interpreting a LEM. That is,
+/// it contains the input, the output and all the assignments resulting from
+/// running one iteration as a HashMap of meta pointers to pointers.
+///
+/// Finally, `visits` contains the data collected from visiting the slots. This
+/// information is used to generte the witness.
+#[derive(Clone)]
+pub struct Frame<F: LurkField> {
+    pub input: [Ptr<F>; 3],
+    pub output: [Ptr<F>; 3],
+    pub binds: HashMap<MetaPtr, Ptr<F>>,
```

Comment:
Would renaming `binds` to `bindings` be inaccurate? This would be much less confusing to me, if so.

Reply:
Will do